### PR TITLE
Add builders for percentile and median accumulators/window functions

### PR DESCRIPTION
**Pull request by vbabanin**
This pull request (PR) introduces the capability to calculate the median and percentile of numeric values in the MongoDB aggregation pipeline for $group and $setWindowFields stages

[JAVA-3860](https://jira.mongodb.org/browse/JAVA-3860)

### CODE REVIEW

File: driver-core/src/main/com/mongodb/client/model/Accumulators.java [hunk 1]
```diff
@@ -67,6 +67,42 @@ public static <TExpression> BsonField avg(final String fieldName, final TExpress
...
+     * @param inExpression   The input expression.
+     * @param pExpression    The expression representing a percentiles of interest.
+     * @param method         The method to be used for computing the percentiles.
+     * @param <InExpression> The type of the input expression.
+     * @param <PExpression>  The type of the percentile expression.
+     * @return The requested {@link BsonField}.
+     * @mongodb.driver.manual reference/operator/aggregation/percentile/ $percentile
+     * @mongodb.server.release 7.0
+     */
+    public static <InExpression, PExpression> BsonField percentile(final String fieldName, final InExpression inExpression,
```
Comment (vbabanin) [inline driver-core/src/main/com/mongodb/client/model/Accumulators.java:85]: In the design doc, you may notice that if no method value is specified to calculate quantiles, the default value is 'discrete', suggesting the possibility of creating an overloaded method without the method argument. However, at present, the decision regarding the default choice for the method has been deferred. Therefore, all methods currently require the method argument.

File: driver-core/src/main/com/mongodb/client/model/Accumulators.java [hunk 2]
```diff
@@ -67,6 +67,42 @@ public static <TExpression> BsonField avg(final String fieldName, final TExpress
...
+     * of percentiles of interest within a group, where each element is a numeric value between 0.0 and 1.0 (inclusive).
+     *
+     * @param fieldName      The field computed by the accumulator.
+     * @param inExpression   The input expression.
+     * @param pExpression    The expression representing a percentiles of interest.
+     * @param method         The method to be used for computing the percentiles.
+     * @param <InExpression> The type of the input expression.
+     * @param <PExpression>  The type of the percentile expression.
+     * @return The requested {@link BsonField}.
+     * @mongodb.driver.manual reference/operator/aggregation/percentile/ $percentile
```
Comment (vbabanin) [inline driver-core/src/main/com/mongodb/client/model/Accumulators.java:82]: Here is the comment from PR 1130, written by @jyemin:

> Add @since 4.10 to both methods

This comment is duplicated from [PR 1130](https://github.com/mongodb/mongo-java-driver/pull/1130#discussion_r1213069726)."

Comment (vbabanin) [reply driver-core/src/main/com/mongodb/client/model/Accumulators.java:82]: Done

File: driver-core/src/test/functional/com/mongodb/client/model/AggregatesFunctionalSpecification.groovy [hunk 3]
```diff
@@ -297,6 +299,42 @@ class AggregatesFunctionalSpecification extends OperationFunctionalSpecification
         results.first().get('res', List) == ['c']
     }
 
+    @IgnoreIf({ serverVersionLessThan(7, 0) })
```
Comment (katcharov) [inline driver-core/src/test/functional/com/mongodb/client/model/AggregatesFunctionalSpecification.groovy:302]: While testing $unset and $geoNear, I added `AggregatesTest`. I preferred this because I found groovy hard to follow, and because it did not show the code as the user would see/write it. Should we write all new tests in Java? (I understand these tests are already written, like the existing tests.)

Comment (vbabanin) [reply driver-core/src/test/functional/com/mongodb/client/model/AggregatesFunctionalSpecification.groovy:302]: Thank you for sharing your perspective. Personally, I find Groovy tests to be descriptive and convenient in certain scenarios. Java tests can occasionally become bulky. However, I agree with your viewpoint that writing tests in Java brings us closer to the user experience. I will make the change and switch them to Java.

Comment (vbabanin) [reply driver-core/src/test/functional/com/mongodb/client/model/AggregatesFunctionalSpecification.groovy:302]: Thank you for sharing your perspective. Personally, I find Groovy tests to be descriptive and convenient for data-driven tests. While Java tests can occasionally become bulky, I agree with your viewpoint that writing tests in Java brings us closer to the user experience. At the same time, I believe that we could still leverage vanilla Java within Groovy to achieve that goal and make test preparation and assertions with Groovy syntax.

Taking everything into consideration, I think it's best to stick with our usual practice of writing new tests in Java and avoid adding more Groovy tests to the specifications. So, I'll make the switch to Java for new test cases.

File: driver-core/src/test/unit/com/mongodb/client/model/TestWindowOutputFields.java [hunk 4]
```diff
@@ -212,6 +222,111 @@ void pick() {
...
+        assertThrows(IllegalArgumentException.class, () -> WindowOutputFields.percentile(PATH, null, pExpressionParameter, method, window), msg);
+        assertThrows(IllegalArgumentException.class, () -> WindowOutputFields.percentile(PATH, inExpressionParameter, null, method, window), msg);
+        assertThrows(IllegalArgumentException.class, () -> WindowOutputFields.percentile(PATH, inExpressionParameter, pExpressionParameter, null, window), msg);
+    }
+
+    @ParameterizedTest
+    @MethodSource("medianWindowFunctionsSource")
+    void median(final Object inExpressionParameter,
+                final BsonValue expectedInExpression,
+                final Window window) {
```
Comment (stIncMale) [inline driver-core/src/test/unit/com/mongodb/client/model/TestWindowOutputFields.java:256]:
```suggestion
                @NUllable final Window window) {
```

**Review by stIncMale — CHANGES_REQUESTED** [review]

File: driver-core/src/main/com/mongodb/client/model/Accumulators.java [hunk 5]
```diff
@@ -67,6 +67,44 @@ public static <TExpression> BsonField avg(final String fieldName, final TExpress
...
+     * @since 4.10
+     * @mongodb.server.release 7.0
+     */
+    public static <InExpression, PExpression> BsonField percentile(final String fieldName, final InExpression inExpression,
+                                                                   final PExpression pExpression, final String method) {
```
[the remark this answers is not in the document]
Comment (stIncMale) [reply driver-core/src/main/com/mongodb/client/model/Accumulators.java:87]: The PR I mentioned in footnote 1: https://github.com/mongodb/mongo-java-driver/pull/1141

File: driver-core/src/main/com/mongodb/client/model/Accumulators.java [hunk 6]
```diff
@@ -67,6 +67,44 @@ public static <TExpression> BsonField avg(final String fieldName, final TExpress
...
+     * @param pExpression    The expression representing a percentiles of interest.
+     * @param method         The method to be used for computing the percentiles.
+     * @param <InExpression> The type of the input expression.
+     * @param <PExpression>  The type of the percentile expression.
+     * @return The requested {@link BsonField}.
+     * @mongodb.driver.manual reference/operator/aggregation/percentile/ $percentile
+     * @since 4.10
+     * @mongodb.server.release 7.0
+     */
+    public static <InExpression, PExpression> BsonField percentile(final String fieldName, final InExpression inExpression,
```
[the remark this answers is not in the document]
Comment (vbabanin) [reply driver-core/src/main/com/mongodb/client/model/Accumulators.java:86]: Currently, not all array expressions are supported in MQL.

Additionally, if we start incorporating MQLValue explicitly into the Accumulators builder class for new methods, it would introduce a combination of the old and new approaches. Some methods would accept Object/Bson, while others would accept MQL. This could potentially confuse users, especially those who are still using the Document/BsonDocument approach in their code.

In my opinion, it would be better to wait until aggregations are available in the computational API, where we could use only MQL entities.

[the remark this answers is not in the document]
Comment (vbabanin) [reply driver-core/src/main/com/mongodb/client/model/Accumulators.java:86]: As per our discussion today, I have created a separate ticket for supporting these accumulators in the $project stage. You can find the ticket at
[JAVA-5025](https://jira.mongodb.org/browse/JAVA-5025)

File: driver-core/src/main/com/mongodb/client/model/ApproximateQuantileMethod.java [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+ * limitations under the License.
+ */
+package com.mongodb.client.model;
+
+import com.mongodb.annotations.Sealed;
+
+
+/**
+ * @see QuantileMethod#approximate()
+ * @since 4.10
```
Comment (stIncMale) [inline driver-core/src/main/com/mongodb/client/model/ApproximateQuantileMethod.java:23]: Let's add `@mongodb.server.release 7.0`.

File: driver-core/src/test/functional/com/mongodb/client/model/OperationTest.java [hunk 8]
```diff
@@ -114,7 +114,6 @@ protected void assertResults(final List<Bson> pipeline, final String expectedRes
     protected List<Object> aggregateWithWindowFields(@Nullable final Object partitionBy, final WindowOutputField output) {
         List<Bson> stages = new ArrayList<>();
         stages.add(setWindowFields(partitionBy, null, output));
-        stages.add(sort(ascending("num1")));
```
Comment (stIncMale) [inline driver-core/src/test/functional/com/mongodb/client/model/OperationTest.java]: The sorting is important here:

- The order of documents produced by `$setWindowFields` is not guaranteed, even if `sortBy` for `$setWindowFields` is specified, as it is used only to within the stage.
- The test compares two lists, i.e., the test relies on the order.

Clearly, sorting by `num1` can't be hardcoded in a method that is located in `OperationTest`, but this method may accept a `@Nullable Bson` sort specification, allowing the test `shouldSetWindowFieldWithQuantile` to specify the sorting it needs.

P.S. Note that this is not a problem in `AggregatesTest.shouldGroupWithQuantile`, because there you filter the output to get the single document you need by its `_id`.

**Review by stIncMale — CHANGES_REQUESTED** [review]

**Review by stIncMale — APPROVED** [review]

File: driver-core/src/test/functional/com/mongodb/client/model/AggregatesTest.java [hunk 9]
```diff
@@ -18,26 +18,120 @@
...
+                .filter(document -> document.get("_id").equals(new Document("gid", true)))
+                .findFirst().map(document -> document.get("sat_95")).get();
 
+        assertEquals(expectedGroup1, result);
+
+        result = results.stream()
+                .filter(document -> document.get("_id").equals(new Document("gid", false)))
+                .findFirst().map(document -> document.get("sat_95")).get();
+
+        assertEquals(expectedGroup2, result);
```
Comment (katcharov) [inline driver-core/src/test/functional/com/mongodb/client/model/AggregatesTest.java:94]: It seems that the above boilerplate could be replaced with, for example:
```
        assertResults(pipeline, "[\n"
                + "   { _id: true, sat_95: [3, 2] },\n"
                + "   { _id: false, sat_95: [1, 2] }\n"
                + "]");
```
This seems easier to follow than using a stream to pick out the ids, and then manually matching on the particular field.

So instead of the ~10 lines of boilerplate, plus these ~3 lines per test:

```
Arguments.of(
    percentile("sat_95", "$x", new double[]{0.95}, QuantileMethod.approximate()),
    asList(3.0), asList(1.0)),
```

there could be ~6 lines per test:

```
        List<Bson> pipeline = assertPipeline(
                ...
                group("$z", percentile("sat_95", "$x", ofNumberArray(0.95))));

        assertResults(pipeline, "[\n"
                + "   { _id: true, sat_95: [3, 2] },\n"
                + "   { _id: false, sat_95: [1, 1] }\n"
                + "]");
```

I count the lines because parameterization is often used to decrease boilerplate and line count, but to me it seems there can often be other ways to deal with that problem which ultimately make the code more clear.

I think it is fine to have a some duplication in the "data" part of a test (for example, trying to remove it separates `percentile` from `group`, and the array values from `sat_95`, which makes things harder to follow).

**Review by katcharov — CHANGES_REQUESTED** [review]
(This is a partial review.)

File: driver-core/src/test/functional/com/mongodb/client/model/AggregatesTest.java [hunk 10]
```diff
@@ -18,26 +18,120 @@
...
+import static org.hamcrest.Matchers.hasSize;
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assumptions.assumeTrue;
 
 public class AggregatesTest extends OperationTest {
 
+    private static Stream<Arguments> groupWithQuantileSource() {
+        return Stream.of(
+                Arguments.of(percentile("sat_95", "$x", new double[]{0.95}, QuantileMethod.approximate()), asList(3.0), asList(1.0)),
+                Arguments.of(percentile("sat_95", "$x", new double[]{0.95, 0.3}, QuantileMethod.approximate()), asList(3.0, 2.0),
```
[the remark this answers is not in the document]
Comment (vbabanin) [reply driver-core/src/test/functional/com/mongodb/client/model/AggregatesTest.java:57]: I think we should test whatever the user gives us. They can simply provide an array of doubles, which seems like a reasonable choice. To cover all possible scenarios, I can add` MqlValues.ofNumberArray` and `BsonArray` to the test data. Let me know your thoughts

File: driver-core/src/main/com/mongodb/client/model/Accumulators.java [hunk 11]
```diff
@@ -67,6 +67,51 @@ public static <TExpression> BsonField avg(final String fieldName, final TExpress
...
+     * @param fieldName The field computed by the accumulator.
+     * @param inExpression The input expression.
+     * @param method The method to be used for computing the median.
+     * @param <InExpression> The type of the input expression.
+     * @return The requested {@link BsonField}.
+     * @mongodb.driver.manual reference/operator/aggregation/median/ $median
+     * @since 4.10
+     * @mongodb.server.release 7.0
+     */
+    public static <InExpression> BsonField median(final String fieldName, final InExpression inExpression,  final QuantileMethod method) {
```
[the remark this answers is not in the document]
Comment (vbabanin) [reply driver-core/src/main/com/mongodb/client/model/Accumulators.java:108]: The parameter name 'InExpression' is used in five other methods as well. This naming choice was made to maintain consistency within the documentation. The usage of the type parameter name `'* @param <InExpression> The type of the input expression.'` follows a consistent approach throughout the entire class. If we want to change this approach here, it would require renaming it in all instances. What are your thoughts on whether we should change it for the entire class?

File: driver-core/src/main/com/mongodb/client/model/Accumulators.java [hunk 12]
```diff
@@ -67,6 +67,51 @@ public static <TExpression> BsonField avg(final String fieldName, final TExpress
...
+    }
+
+    /**
+     * Returns a combination of a computed field and an accumulator that generates a BSON {@link org.bson.BsonType#DOUBLE Double }
+     * representing the median value computed from the given {@code inExpression} within a group.
+     *
+     * @param fieldName The field computed by the accumulator.
+     * @param inExpression The input expression.
+     * @param method The method to be used for computing the median.
+     * @param <InExpression> The type of the input expression.
```
[the remark this answers is not in the document]
Comment (vbabanin) [reply driver-core/src/main/com/mongodb/client/model/Accumulators.java:102]: I think it is a good idea. However, i wonder should we mention that on a class level documentation? `MqlValue` can be provided to almost all methods with expressions in `Accumulator` class

File: driver-core/src/main/com/mongodb/client/model/Accumulators.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (katcharov) [reply driver-core/src/main/com/mongodb/client/model/Accumulators.java:108]: Fine to leave

File: driver-core/src/main/com/mongodb/client/model/Accumulators.java (hunk 1 above)
Comment (katcharov) [reply driver-core/src/main/com/mongodb/client/model/Accumulators.java:86]: Discussed, server plans to add new quantile methods

File: driver-core/src/main/com/mongodb/client/model/Accumulators.java [hunk 13]
```diff
@@ -67,6 +67,51 @@ public static <TExpression> BsonField avg(final String fieldName, final TExpress
...
+        notNull("fieldName", fieldName);
+        notNull("inExpression", inExpression);
+        notNull("pExpression", inExpression);
+        notNull("method", method);
+        return quantileAccumulator("$percentile", fieldName, inExpression, pExpression, method);
+    }
+
+    /**
+     * Returns a combination of a computed field and an accumulator that generates a BSON {@link org.bson.BsonType#DOUBLE Double }
+     * representing the median value computed from the given {@code inExpression} within a group.
```
[the remark this answers is not in the document]
Comment (katcharov) [reply driver-core/src/main/com/mongodb/client/model/Accumulators.java:97]: We can address this later, in a future PR (or, decide not to).

File: driver-core/src/main/com/mongodb/client/model/Accumulators.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (katcharov) [reply driver-core/src/main/com/mongodb/client/model/Accumulators.java:102]: This is fine, this would be inconsistent with the rest of the class.

Discussed: we can address this later, in a future PR (or, decide not to).

File: driver-core/src/main/com/mongodb/client/model/QuantileMethod.java [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+     * though they may not be {@linkplain Object#equals(Object) equal}.
+     * <pre>{@code
+     *  QuantileMethod method1 = QuantileMethod.approximate();
+     *  QuantileMethod method2 = QuantileMethod.of(new BsonString("approximate"));
+     * }</pre>
+     *
+     * @param method A {@link BsonValue} representing the required {@link QuantileMethod}.
+     * @return The requested {@link QuantileMethod}.
+     */
+    static QuantileMethod of(final BsonValue method) {
```
[the remark this answers is not in the document]
Comment (katcharov) [reply driver-core/src/main/com/mongodb/client/model/QuantileMethod.java:65]: No: this might be a bson document in the future.

File: driver-core/src/main/com/mongodb/client/model/QuantileMethod.java [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+ *
+ * @see Accumulators#percentile(String, Object, Object, QuantileMethod)
+ * @see Accumulators#median(String, Object, QuantileMethod)
+ * @see WindowOutputFields#percentile(String, Object, Object, QuantileMethod, Window)
+ * @see WindowOutputFields#median(String, Object, QuantileMethod, Window)
+ * @since 4.10
+ * @mongodb.server.release 7.0
+ */
+@Sealed
+public interface QuantileMethod {
```
[the remark this answers is not in the document]
Comment (katcharov) [reply driver-core/src/main/com/mongodb/client/model/QuantileMethod.java:39]: We discussed this. This is at least consistent with what was done for densify and search. There are still some issues that might be relevant in the future, but we should discuss them as they come up, probably as a team.

File: driver-core/src/test/functional/com/mongodb/client/model/AggregatesTest.java [hunk 16]
```diff
@@ -18,26 +18,120 @@
...
+                new Document("x", 2).append("z", true),
+                new Document("x", 3).append("z", true)
+        };
+        getCollectionHelper().insertDocuments(original);
+
+        //when
+        List<Document> results = getCollectionHelper().aggregate(Collections.singletonList(
+                group(new Document("gid", "$z"), quantileAccumulator)), DOCUMENT_DECODER);
+
+        //then
```
[the remark this answers is not in the document]
Comment (katcharov) [reply driver-core/src/test/functional/com/mongodb/client/model/AggregatesTest.java:81]: Discussed: we can address this later, in a future PR (or, decide not to).

File: driver-core/src/test/functional/com/mongodb/client/model/AggregatesTest.java (hunk 9 above)
Comment (katcharov) [reply driver-core/src/test/functional/com/mongodb/client/model/AggregatesTest.java:94]: Discussed: we can address this later, in a future PR (or, decide not to). We should have a consistent style in the tests.

File: driver-core/src/test/functional/com/mongodb/client/model/AggregatesTest.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (katcharov) [reply driver-core/src/test/functional/com/mongodb/client/model/AggregatesTest.java]: Discussed: this should not be passed, we can pass something else.

File: driver-core/src/test/functional/com/mongodb/client/model/AggregatesTest.java [hunk 17]
```diff
@@ -18,26 +18,120 @@
...
+        Document[] original = new Document[]{
+                new Document("x", 1).append("z", false),
+                new Document("x", 2).append("z", true),
+                new Document("x", 3).append("z", true)
+        };
+        getCollectionHelper().insertDocuments(original);
+
+        //when
+        List<Document> results = getCollectionHelper().aggregate(Collections.singletonList(
+                group(new Document("gid", "$z"), quantileAccumulator)), DOCUMENT_DECODER);
```
[the remark this answers is not in the document]
Comment (katcharov) [reply driver-core/src/test/functional/com/mongodb/client/model/AggregatesTest.java]: Discussed: we can address this (DOCUMENT_DECODER) later, in a future PR. ($z fixed)

**Review by katcharov — APPROVED** [review]
LGTM
